### ISSUE TITLE
Possible to use another IOS device as VRF proxy?

### ISSUE THREAD

**Issue by ecm75** [ytti/oxidized#639]
I have the following challange while using oxidized:

I have a lot of IOS router which are not reachable over public internet, however I can reach them through a central IOS souter and reach the private router using the following command on the router CLI:
telnet A.B.C.D vrf VRF_TAG

Would it be possible to implement a model which does this:
- connect to router A, 
- login to router A, 
(- do enable on router A, )
- do telnet ... vrf ... to router B, 
- login to router B,
(do enable on router B,)
- save config of router B, 
- disconnect from router B,
- disconnect from router A

I'd like to achive this by setting additional vars for router B, like eg.
vrf_tag, vrf_host, vrf_user, vrf_pass, vrf_enable

My ruby knowledge is quite limited, so I don't even know where to start...

**Comment by ytti** [comment]
We support SSH proxy right now, I've never tested if IOS router can act as SSH proxy, I don't have high hopes, but worth a shot.

If not, then easiest would be to install server to act as your SSH proxy between the networks. 

My suspicion is that this is highly atypical scenario, and while it wouldn't be hard to add support for this, it would be decent amount of work and refactoring. The complexity would be justifiable if were common requirement, but right now I don't think it is.

**Comment by ecm75** [comment]
I will try with ssh_proxy option, thanks for the fast answer!

**Comment by ecm75** [comment]
Is is possiböle to use ssh_proxy with telnet to the router?
oxidized-host --SSH--> ssh_proxy-host --TELNET--> router

**Comment by ytti** [comment]
Not right now, you could try to change this: https://github.com/ytti/oxidized/blob/master/lib/oxidized/input/ssh.rb#L28 to telnet and tell if it works. 

If it does, I can make it configurable.

**Comment by ecm75** [comment]
No, it doesn't seem to work, but i I understand the code correct it executes something like this on the oxidized host:
`ssh <ssh_proxy> -W <node_host>:<node_port>`
However I can connect to the router using this command on the shell from the oxidized host:
`ssh <ssh_proxy> -W <node_host>:23`
I then get the router login prompt, but even if i change the code in ssh.rb to 
`Net::SSH::Proxy::Command.new("ssh #{proxy_host} -W %h:23)`
it doen't work from within oxidized...

**Comment by ytti** [comment]
I guess you're out of luck then and have to use SSH :/

**Comment by ecm75** [comment]
I solved my problem... but it requires an additional gem and is probably quite ugly (I'm not a developer) but maybe it's of use for someone else.

`gem install net-ssh-gateway`

`diff --git a/lib/oxidized/input/telnet.rb b/lib/oxidized/input/telnet.rb
index a5561b9..6d03d32 100644
--- a/lib/oxidized/input/telnet.rb
+++ b/lib/oxidized/input/telnet.rb
@@ -1,5 +1,6 @@
 module Oxidized
   require 'net/telnet'
+  require 'net/ssh/gateway'
   require 'oxidized/input/cli'
   class Telnet < Input
     RescueFail = {}
@@ -10,13 +11,28 @@ module Oxidized
       @node    = node
       @timeout = Oxidized.config.timeout
       @node.model.cfg['telnet'].each { |cb| instance_exec(&cb) }
-      port = vars(:telnet_port) || 23
-
-      opt = { 'Host'    => @node.ip,
+      if proxy_host = vars(:ssh_proxy)
+        Oxidized.logger.debug "Telnet: using SSH proxy #{proxy_host}"
+        port = vars(:telnet_port) || 23
+        gateway = Net::SSH::Gateway.new(proxy_host,
+           'oxidized',
+           :verbose => :error)
+        
+        local_port = gateway.open(@node.ip,port)
+
+        opt = { 'Host'    => "127.0.0.1",
+              'Port'    => local_port.to_i,
+              'Timeout' => @timeout,
+              'Model'   => @node.model }
+        opt['Output_log'] = Oxidized::Config::Log + "/#{@node.ip}-telnet" if Oxidized.config.input.debug?
+      else
+        port = vars(:telnet_port) || 23
+        opt = { 'Host'    => @node.ip,
               'Port'    => port.to_i,
               'Timeout' => @timeout,
               'Model'   => @node.model }
-      opt['Output_log'] = Oxidized::Config::Log + "/#{@node.ip}-telnet" if Oxidized.config.input.debug?
+        opt['Output_log'] = Oxidized::Config::Log + "/#{@node.ip}-telnet" if Oxidized.config.input.debug?
+      end
 
       @telnet  = Net::Telnet.new opt
       if @node.auth[:username] and @node.auth[:username].length > 0
@@ -62,6 +78,10 @@ module Oxidized
         disconnect_cli
         @telnet.close
       rescue Errno::ECONNRESET
+        if gateway
+          gateway.close(port)
+          gateway.shutdown!
+        end
       end
     end`

**Comment by SjonsDePlons** [comment]
I'm a bit late into the discussion, but @ecm75 can you clarify what you did in your last post? I have almost the same problem but not with telnet but with ssh.
